Single Sign-On Authentication Mechanism

TheMysteriousX opened the pull request:
This is an authentication module that can be used to integrate LibreNMS with single sign-on providers.

The authentication code is imperative rather than object oriented; this has meant that I've had to make more changes than I'd like.

There are tests included, but they must be executed standalone and with config.php manually set (hence the .disabled extension) - the authentication code being imperative means that once a mechanism is loaded, it cannot be unloaded or replaced, breaking other tests.

I've also stopped http-auth being special cased in authenticate.php by tweaking it to use the two methods I've added.

It's been a few years since I've written any PHP, so feedback is gratefully received.

DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/7add769e-3480-44b5-91aa-68c10ce0caee) was created.

TheMysteriousX commented:
Looks like there are a few compatibility problems to fix up - I'll sort those momentarily.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/cac9f93e-79a3-41dc-aaee-cb7a57be47a6) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/acb10988-817e-4850-a85d-ad167dd6e9ef) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/cac9f93e-79a3-41dc-aaee-cb7a57be47a6) completed: **40 new issues, 8 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/acb10988-817e-4850-a85d-ad167dd6e9ef) completed: **40 new issues, 8 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/8adf46f6-bfd1-40ba-9558-cd901cec2edd) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/125f0b66-7a2c-4b2d-b52d-75ba557b4b4a) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/125f0b66-7a2c-4b2d-b52d-75ba557b4b4a) completed: **40 new issues, 8 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/91cebb06-5de6-4097-8b1a-9be4a6ee004c) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/91cebb06-5de6-4097-8b1a-9be4a6ee004c) completed: **40 new issues, 8 updated code elements**

TheMysteriousX commented:
Code now passes Travis' checks under PHP 5.3.
I can't do much about the Scrutinizer failures; what it's flagging up is 100% correct, but is due to the auth mechanisms having repeated code and identical functions in each file.

laf commented:
Thanks for contributing this.

I've not gone through the code in detail yet and have no real way to test this but it seems like it would be better to put the config into another level within the config array like radius: https://docs.librenms.org/#Extensions/Authentication/#radius-authentication

I.e `$config['sso']['sso_group_strategy']`

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/479da396-ddc8-4510-b320-7c2768e14a09) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/fdef9909-1cfa-4769-be4b-6d0fdee41b9d) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/479da396-ddc8-4510-b320-7c2768e14a09) completed: **40 new issues, 8 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/fdef9909-1cfa-4769-be4b-6d0fdee41b9d) completed: **40 new issues, 8 updated code elements**

TheMysteriousX commented:
Just pushed a new version with the configuration moved down a level.

I've not tested it this way, but it should be possible to simulate a single sign-on environment using SetEnv - something like this in your Apache config:

```
SetEnv REMOTE_USER ssouser
SetEnv mail ssouser@example.org
SetEnv displayName "SSO User"
SetEnv level 10
```

Then a basic config like:

```
$config['sso']['mode'] = 'env';
$config['sso']['realname_attr'] = 'displayName';
$config['sso']['email_attr'] = 'mail';
$config['sso']['level_attr'] = 'level';
$config['sso']['create_users'] = true;
$config['sso']['update_users'] = true;
$config['sso']['group_strategy'] = 'attribute';
```

Which should create an entry in the db for 'ssouser', and allow access at level 10. Might be a few rough edges in this configuration, but it should work.

laf commented on html/includes/authentication/sso.inc.php:
```diff
@@ -0,0 +1,336 @@
+use LibreNMS\Exceptions\InvalidIpException;
+
+/**
+ * Some functionality in this mechanism is inspired by confluence_http_authenticator (@chauth) and graylog-plugin-auth-sso (@Graylog)
```
Please add both of these packages to the Acknowledgements doc: https://github.com/librenms/librenms/blob/master/doc/General/Acknowledgement.md

TheMysteriousX replied:
Done.

laf commented on html/includes/authentication/sso.inc.php:
```diff
@@ -0,0 +1,336 @@
+function authenticate($username, $password)
+{
+    if (empty($username)) {
+        throw new AuthenticationException('sso.user_attr was not found or was empty');
```
Users don't see config values like that so you will need to `$config['sso']['user_attr']` in this output

TheMysteriousX replied:
Done.

laf commented on html/includes/authentication/sso.inc.php:
```diff
@@ -0,0 +1,336 @@
+    $level = auth_sso_calculate_level();
+
+    // User has already been approved by the authenicator so if automatic user create/update is enabled, do it
+    if (Config::get('sso.create_users') && !user_exists($username)) {
```
Is this block going to be called on every page and every graph load as we run `authenticate()` each time.

TheMysteriousX replied:
I don't believe that's the case - I threw in a print statement and authenticate() only seems to be called on the first request. It gets short circuited if there's an active session.

If it did get called repeatedly (user with broken cookie settings) it's idempotent, and if a database cache is present, the impact is just an additional sql query per page load.

laf replied:
Actually I've just remember we switched to using sessions once validated so you're right on this.

laf commented on html/includes/print-menubar.php:
```diff
@@ -639,7 +639,7 @@
 if ($_SESSION['authenticated']) {
     echo('
-           <li><a href="logout/"><i class="fa fa-sign-out fa-fw fa-lg" aria-hidden="true"></i> Logout</a></li>');
+           <li><a href="' . $config['auth_logout_handler']  . '"><i class="fa fa-sign-out fa-fw fa-lg" aria-hidden="true"></i> Logout</a></li>');
```
Should switch to `Config::get()` here as well.

TheMysteriousX replied:
Done, by changing how this is done. The template isn't really the right place to handle authentication stuff as it's prone to being altered. I implemented the same functionality by making the URL the user is sent to after log out configurable.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/42731263-7781-40ca-9251-e548b7e8ce15) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/42731263-7781-40ca-9251-e548b7e8ce15) completed: **40 new issues, 8 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/34d1eac2-b8cb-44fc-8040-dff390c9152e) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/34d1eac2-b8cb-44fc-8040-dff390c9152e) completed: **40 new issues, 8 updated code elements**

laf commented:
LGTM.

@librenms/reviewers anyone else?

laf commented:
one thing to note actually. Someone has got a PR in the move all of the auth modules to classes #7497 

As that's been in for a while, it would be unfair to ask that user to do the work to convert this as well. Would you be able to migrate this once 7497 is merged?

TheMysteriousX commented:
Sure that's fine - I have a couple of questions about the new structure, I'll pop them in the other pull request.

TheMysteriousX commented:
Answered my questions with a bit more reading.

laf commented:
@TheMysteriousX The other PR is now merged in.

TheMysteriousX commented:
Naïve rebase, do not merge as it probably doesn't work (yet).

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/937fd10c-cce7-48a7-b06c-931331168ade) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/937fd10c-cce7-48a7-b06c-931331168ade) completed: **9 new issues, 37 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/d70ea4e3-eca8-4867-8acd-7ebc8043fa68) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/d70ea4e3-eca8-4867-8acd-7ebc8043fa68) completed: **5 new issues, 33 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/dfa6adc5-908d-428c-95a4-b5b65dd80b6e) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/dfa6adc5-908d-428c-95a4-b5b65dd80b6e) completed: **18 new issues, 23 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/f1e4c1e6-7bf4-42bb-82a4-961068557ef7) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/f1e4c1e6-7bf4-42bb-82a4-961068557ef7) completed: **11 new issues, 22 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/a20f7f59-ac6b-44e1-b38f-c49328d70301) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/a20f7f59-ac6b-44e1-b38f-c49328d70301) completed: **11 new issues, 42 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/5ff67eca-9a2a-46c7-86fb-13cf99234529) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/5ff67eca-9a2a-46c7-86fb-13cf99234529) completed: **11 new issues, 43 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/5028b9cd-52ce-4a8f-b4d5-420a341db298) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/5028b9cd-52ce-4a8f-b4d5-420a341db298) completed: **10 new issues, 24 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/bebf4440-7906-40cd-a036-f02bc4be64aa) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/bebf4440-7906-40cd-a036-f02bc4be64aa) completed: **11 new issues, 43 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/af7ce422-58cf-4af6-ba4b-8f9d1e238d91) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/af7ce422-58cf-4af6-ba4b-8f9d1e238d91) completed: **11 new issues, 43 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/ddf752d6-c04d-4769-8205-65cb6159ffdf) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/ddf752d6-c04d-4769-8205-65cb6159ffdf) completed: **9 new issues, 43 updated code elements**

TheMysteriousX commented:
I think this is now ported. I'll give it another once over with fresher eyes tomorrow.

Summary of changes:
* Migrated a couple of methods into the interface/base class
* Converted imperative mechanism to class based
* Enabled tests as authentication is now mutable
* Added Auth::reset to allow a new mechanism to be loaded at runtime

The CI check that is failing is PHP5.3 - I've used the wrong array syntax to support this.

I can go back and change the syntax, but I think PHP5.3 was being dropped?

laf commented:
It's not being dropped until January so this will need to be resolved before then.

murrant commented on LibreNMS/Authentication/AuthorizerBase.php:
```diff
@@ -244,4 +245,14 @@ public function updateUser($user_id, $realname, $level, $can_modify_passwd, $ema
         return 0;
     }
+
+    public function authExternal()
```
This would probably be better named `authIsExternal()`

murrant commented:
:+1: for implementing unit tests :)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/6e6a8934-c0fe-4331-8e3d-19d575a01234) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/6e6a8934-c0fe-4331-8e3d-19d575a01234) completed: **9 new issues, 43 updated code elements**

TheMysteriousX commented:
Great, thanks for reviewing.

TheMysteriousX commented:
Thanks for merging - the company I work for use this mechanism on their instance, so feel free to @ me if there are any issues.

murrant commented:
@TheMysteriousX I'm looking to refactor this a bit to inheret from MysqlAuthorizer.

I can't figure out why you ignore the given $username in functions and use $this->getExternalUsername() instead?  Would this cause errors if we query for another user besides the logged in one?

TheMysteriousX commented:
That would be fine.

The username passed by functions should always be taken by from getExternalUsername, so by taking it from there it eliminates the possibility that the username could have been modified (accidentally or maliciously).

murrant commented:
But those functions return info about the requested user, not the logged in user. 

For example, say you are viewing a shared dashboard with your code all shared dashboards would show the currently logged in user as the owner.

TheMysteriousX commented:
Ok, I misunderstood their purpose then - I assumed getUserid/getUserLevel/userExists were intended as internal methods as they're not called outside of the auth code (except for scripts that do weird things, like the IRC bot).

I've checked and it doesn't seem to affect functionality - dashboards are all correctly owned, modifying users doesn't result in incorrect values being assigned etc.

I can send a pull to change the values if you like, or have you already changed it in your branch?

murrant commented:
The plan is to change the class to extend MysqlAuth (to remove code duplication) and remove extraneous functions.  I would prefer if you make the change as I can only test with the unit tests.
